pkg/controller/vitessbackupschedule/vitessbackupschedule_controller.go

```diff
@@ -0,0 +1,634 @@
+			// Return and don't requeue
+			return resultBuilder.Result()
+		}
+		// Error reading the object - requeue the request.
+		return resultBuilder.Error(err)
+	}
+
+	// Register this reconciling attempt no matter if we fail or succeed.
+	defer func() {
+		reconcileCount.WithLabelValues(vbsc.Name, metrics.Result(err)).Inc()
+	}()
+
+	jobs, mostRecentTime, err := r.getJobsList(ctx, req, vbsc.Name)
+	if err != nil {
+		// We had an error reading the jobs, we can requeue.
+		return resultBuilder.Error(err)
+	}
+
+	err = r.updateVitessBackupScheduleStatus(ctx, mostRecentTime, vbsc, jobs.active)
+	if err != nil {
+		// We had an error updating the status, we can requeue.
+		return resultBuilder.Error(err)
+	}
+
+	// We must clean up old jobs to not overcrowd the number of Pods and Jobs in the cluster.
+	// This will be done according to both failedJobsHistoryLimit and successfulJobsHistoryLimit fields.
+	r.cleanupJobsWithLimit(ctx, jobs.failed, vbsc.GetFailedJobsLimit())
+	r.cleanupJobsWithLimit(ctx, jobs.successful, vbsc.GetSuccessfulJobsLimit())
+
+	err = r.removeTimeoutJobs(ctx, jobs.active, vbsc.Name, vbsc.Spec.JobTimeoutMinutes)
+	if err != nil {
+		// We had an error while removing timed out jobs, we can requeue
+		return resultBuilder.Error(err)
+	}
+
+	// If the Suspend setting is set to true, we can skip adding any job, our work is done here.
+	if vbsc.Spec.Suspend != nil && *vbsc.Spec.Suspend {
+		log.Info("VitessBackupSchedule suspended, skipping")
+		return ctrl.Result{}, nil
+	}
+
+	missedRun, nextRun, err := getNextSchedule(vbsc, time.Now())
+	if err != nil {
+		log.Error(err, "unable to figure out VitessBackupSchedule schedule")
+		// Re-queuing here does not make sense as we have an error with the schedule and the user needs to fix it first.
+		return ctrl.Result{}, nil
+	}
+
+	// Ask kubernetes to re-queue for the next scheduled job, and skip if we don't miss any run.
+	scheduledResult := ctrl.Result{RequeueAfter: nextRun.Sub(time.Now())}
+	if missedRun.IsZero() {
+		return scheduledResult, nil
+	}
+
+	// Check whether we are too late to create this Job or not. The startingDeadlineSeconds field will help us
+	// schedule Jobs that are late.
+	tooLate := false
+	if vbsc.Spec.StartingDeadlineSeconds != nil {
+		tooLate = missedRun.Add(time.Duration(*vbsc.Spec.StartingDeadlineSeconds) * time.Second).Before(time.Now())
+	}
+	if tooLate {
+		log.Info("missed starting deadline for last run, sleeping till next")
+		return scheduledResult, nil
+	}
+
+	// Check concurrency policy and skip this job if we have ForbidConcurrent set plus an active job
+	if vbsc.Spec.ConcurrencyPolicy == planetscalev2.ForbidConcurrent && len(jobs.active) > 0 {
+		log.Infof("concurrency policy blocks concurrent runs: skipping, number of active jobs: %d", len(jobs.active))
+		return scheduledResult, nil
+	}
+
+	// Check concurrency policy to know if we should replace existing jobs
+	if vbsc.Spec.ConcurrencyPolicy == planetscalev2.ReplaceConcurrent {
```

Comment:
has anyone asked for this option? seems like an odd one. i would rather let a backup complete, even if it's taking a long time, than just replace it. if there's a persistent issue that is preventing backups from completing within their scheduled interval, then applying this option will just mean consuming lots of resources and never getting a successful backup.

Reply:
no one has asked for this, i will proceed to remove it

Reply:
fixed via 36d1b40